trie/fog.py

```diff
@@ -0,0 +1,284 @@
+import ast
+from itertools import zip_longest
+from typing import (
+    Any,
+    Sequence,
+    Tuple,
+)
+
+from eth_utils import (
+    to_tuple,
+    ValidationError,
+)
+from sortedcontainers import SortedSet
+
+from trie.typing import (
+    Nibbles,
+    NibblesInput,
+    RawHexaryNode,
+)
+from trie.utils.nodes import (
+    key_starts_with,
+)
+
+
+class HexaryTrieFog:
+    """
+    Keeps track of which parts of a trie have been verified to exist.
+
+    Named after "fog of war" popular in video games like... Red Alert? IDK, I'm old.
+
+    Object is immutable. Any changes, like marking a key prefix as complete, will
+    return a new HexaryTrieFog object.
+    """
+    _unexplored_prefixes: SortedSet
+
+    # INVARIANT: No unexplored prefix may start with another unexplored prefix
+    #   For example, _unexplored_prefixes may not be {(1, 2), (1, 2, 3)}.
+
+    def __init__(self) -> None:
+        # Always start without knowing anything about a trie. The only unexplored
+        #   prefix is the root prefix: (), which means the whole trie is unexplored.
+        self._unexplored_prefixes = SortedSet({()})
+
+    def __repr__(self) -> str:
+        return f"HexaryTrieFog<{self._unexplored_prefixes!r}>"
+
+    @property
+    def is_complete(self) -> bool:
+        return len(self._unexplored_prefixes) == 0
+
+    def explore(
+            self,
+            old_prefix: NibblesInput,
+            sub_segments_input: Sequence[NibblesInput]) -> 'HexaryTrieFog':
```

Comment:
IIUC `sub_segment_input` is: The nibbles extensions from `old_prefix` that are still unexplored.

If so maybe this would benefit from a more descriptive name like `unexplored_sub_segments`  or `sub_segments_still_unexplored` or `foggy_sub_segments`